core/main/src/broker/workflow_broker.rs

```diff
@@ -0,0 +1,316 @@
+                                        .into();
+                                    Self::send_broker_failure_response(&callback, boo);
+                                }
+                            },
+                        }
+                    }
+                    None => {
+                        error!("Failed to receive message");
+                        break;
+                    }
+                }
+            }
+        });
+        BrokerSender { sender: tx }
+    }
+}
+
+impl EndpointBroker for WorkflowBroker {
+    fn get_broker(
+        _request: BrokerConnectRequest,
+        callback: BrokerCallback,
+        broker_state: &mut EndpointBrokerState,
+    ) -> Self {
+        Self {
+            sender: Self::start(callback, broker_state.clone()),
+        }
+    }
+
+    fn get_sender(&self) -> super::endpoint_broker::BrokerSender {
+        self.sender.clone()
+    }
+
+    fn get_cleaner(&self) -> super::endpoint_broker::BrokerCleaner {
+        BrokerCleaner::default()
+    }
+}
+/*
+write exhaustive tests for the WorkflowBroker
+   */
+
+#[cfg(test)]
+pub mod tests {
+
+    use ripple_sdk::{api::gateway::rpc_gateway_api::RpcRequest, tokio, Mockable};
+    use serde_json::json;
+
+    use crate::broker::{
+        endpoint_broker::{BrokerCallback, BrokerRequest, EndpointBrokerState},
+        rules_engine::{JsonDataSource, Rule, RuleEngine},
+    };
+    pub fn broker_request(callback: BrokerCallback) -> BrokerRequest {
+        let mut rule = Rule {
+            alias: "module.method".to_string(),
+            ..Default::default()
+        };
+        let source = JsonDataSource {
+            method: "module.method".to_string(),
+            namespace: Some("module".to_string()),
+            ..Default::default()
+        };
+
+        rule.sources = Some(vec![source]);
+        BrokerRequest {
+            rpc: RpcRequest::mock(),
+            rule,
+            subscription_processed: None,
+            workflow_callback: Some(callback),
+        }
+    }
+    pub fn rule_engine() -> RuleEngine {
+        let engine = RuleEngine::load_from_string_literal(
+            json!({
+                    "endpoints": {
+                        "workflow" : {
+                            "protocol": "workflow",
+                            "url": "http://asdf.com",
+                            "jsonrpc": false
+                          }
+                    },
+                    "rules": {
+                        "static.rule": {
+                            "alias": "static",
+                            "transform": {
+                              "response": "\"Sky\""
+                            }
+                        },
+                        "module.method": {
+                            "alias": "workflow",
+                            "endpoint": "workflow",
+                            "sources": [{
+                                "method": "static.rle"
```

Comment:
there is a typo here for static.rle does it still work is the test for invalid rule?

Reply:
In this case (the unit test test RuleEngine::load_from_string_literal, which makinly tests that it can take a schema compliant string as input, and produce a rule engine from them), so it does not make any difference (but does look a little confusing now that i'm looking at it)